psychopy/experiment/components/_base.py

```diff
@@ -338,50 +338,58 @@ def writeStartTestCode(self, buff, extra=""):
             `%(key)s` syntax to insert the values of any necessary params. Default is an empty 
             string.
         """
+        # create copy of params dict so we can change stuff without harm
+        params = self.params.copy()
 
         # Get starting indent level
         startIndent = buff.indentLevel
 
-        if self.params['startVal'].val in ('', None, -1, 'None'):
-            # If we have no start time, don't write start code
-            return buff.indentLevel - startIndent
+        if params['startVal'].val in ('', None, -1, 'None'):
+            if extra:
+                # if we have extra and no stop condition, extra is the only stop condition
+                params['startType'] = params['startType'].copy()
+                params['startVal'] = params['startVal'].copy()
+                params['startType'].val = "condition"
+                params['startVal'].val = "False"
+            else:
+                # if we just have no stop time, don't write stop code
+                return buff.indentLevel - startIndent
 
         # Newline
         buff.writeIndentedLines("\n")
 
-        if self.params['syncScreenRefresh']:
+        if params['syncScreenRefresh']:
             tCompare = 'tThisFlip'
         else:
             tCompare = 't'
-        params = self.params
         t = tCompare
         # add handy comment
         code = (
             "# if %(name)s is starting this frame...\n"
         )
         buff.writeIndentedLines(code % params)
         # add starting if statement
-        if self.params['startType'].val == 'time (s)':
+        if params['startType'].val == 'time (s)':
             # if startVal is an empty string then set to be 0.0
             if (
-                isinstance(self.params['startVal'].val, str) 
-                and not self.params['startVal'].val.strip()
+                isinstance(params['startVal'].val, str) 
+                and not params['startVal'].val.strip()
             ):
-                self.params['startVal'].val = '0.0'
+                params['startVal'].val = '0.0'
```

Comment:
I'm a little concerned about this line. Originally this is updating the `self.params['startVal']` value, so if anything downstream expects a non-empty `self.params['startVal']`, it would be fine.

Now that we are not modifying in place, could this cause something to fail if something like `if self.params['startVal'] == '0.0'` gets used? I don't know all the places where component objects are used in PsychoPy, so I can't tell for sure. Maybe it's worth generating a few builder scripts after this change to ensure no unexpected bug gets caused by the modification to this line.

Reply:
Good catch, I see what you mean... I think the original behaviour is actually wrong here - otherwise we'd have methods which only execute successfully after the code has been compiled once, which would definitely not be desirable. For that reason I think we've got checks downstream to handle empty startVal. I'll add some tests to the test suite which create Components with blank start/stop fields to make sure this is the case.